### ISSUE TITLE
Guidance on storing money values in SQLite

### ISSUE THREAD

**Issue by Chris-Marassovich** [dotnet/efcore#7232]
I am seeking some guidance for my project which will need to store money amounts in a SQLite DB from a UWP app.

From my research the .NET type of decimal is the preferred type for money.  The rest of my question assumes this is correct.
My model has a type 
```c#
public decimal Amount { get; set; } 
```
and as I follow the document getting started for UWP and SQLite I create a migration.
I note that in the migration it appears to tell SQLite to create a column of type decimal.
My migration has the following:
```c#
Amount = table.Column<decimal>(nullable:False),
```
From my research SQLite has no decimal type and many people appear to prefer the use of an integer and do the old */100 to find the decimal in the integer ... now I am showing my age as that is an old COBOL trick.

I would love to just store the money and issue ef statements to sum the money and I should not care where ef decides to execute those statements (in SQL or client). The Utopia :-)

I am concerned that allowing ef to handle my money storage as it may not be accurate ... but then getting involved and rolling my own storage seems silly if I don't have to.  
Hence here is my question on guidance?
I do not need to care as ef has me covered or I need to roll my own?

If I need to roll my own what is the consensus?  Use the integer and the */100 trick within my models to expose a decimal?

Regards
Chris ....

### Further technical details
Database Provider: SQLite: 1.1.0 Tools: 1.1.0-preview4-final
Operating system: Win 10
IDE: Visual Studio 2015 patch 3

**Comment by bricelam** [comment]
[Microsoft.Data.Sqlite](https://github.com/aspnet/Microsoft.Data.Sqlite) (the ADO.NET provider) will convert the decimals to `TEXT` for lossless storage. This can get a little weird when you perform operations on the columns in queries. See [Datatypes In SQLite Version 3](https://www.sqlite.org/datatype3.html) for all the details. At a high-level, it will do its best to perform a lossless numeric conversion (e.g. to the equivalent of `ulong` or `double`), but it may have to resort to a lossy conversion.

So basically, if all you're worried about is reading and writing the values, you're good. But if you're performing operations on them in your queries, be aware of the corner cases.

**Comment by Chris-Marassovich** [comment]
Thanks @bricelam for such a quick response.
I am still fairly new to this .NET and EF world so I would to just spell out my assumption here and if you could confirm this I would appreciate it.

Using my sample code above if I wrote something like this:
``
using (var db = new myContext())
{
   _TotalValue = db.Transactions.Sum(t => t.Amount);
}
``
Then I assume ef will not issue raw SQL but would bring the data into my classes (where the amount is a decimal) and issue some form of Sum against my classes (ie on the client) to ensure correct values?

This would also be true for OrderBy and other methods ... as in ef will be aware of the storage as text and do all it's work on the client side where the data is a decimal and results will be correct.

I think this is the way it would work which means when the day comes that SQLite can handle money types and the provider is updated then ef will automatically start implementing SQL where appropriate of course.

Thank you for your time.
Regards
Chris ...

**Comment by bricelam** [comment]
`Sum` will be translated to SQL by EF. To ensure it gets evaluated on the "client" (i.e. in .NET), use `.ToList()` for force EF to hit the database.
```C#
    _TotalValue = db.Transactions.ToList().Sum(t => t.Amount);
```

Comparisons are always be lossless on SQLite. When using `.OrderBy()`, the rows will be ordered by their TEXT values.

If SQLite ever adds additional primitive types (they only really have four: `TEXT`, `BLOB`, 'INTEGER` & `REAL`), we will certainly light them up in `Microsoft.Data.Sqlite` and Entity Framework Core.

**Comment by bricelam** [comment]
You can [use Logging](https://docs.microsoft.com/en-us/ef/core/miscellaneous/logging) to see the queries EF is sending to the database.

**Comment by ws4u** [comment]
As mentioned by @bricelam , when using .OrderBy(), the rows will be ordered by their TEXT values on the database.  This produces weird order results for my 'Price' column. Say, $9 will be deemed larger than $13. In my model definition, Price is of type decimal. And in the database, it is converted to TEXT. Do you have a way to make it converted to 'REAL'? Or do you have other solutions to make the sorting happen numerically?

Thanks.

**Comment by bricelam** [comment]
In EF Core 2.2, `OrderBy()` should be evaluated in .NET giving you the expected results (fixed by PR #11889)

**Comment by jonathanantoine** [comment]
Is there a workaround to fix it in a EF 1.X configuration ?

**Comment by bricelam** [comment]
@jonathanantoine Adding `.AsEnumerable()` (or `ToList` or `ToArray`) before the `.OrderBy()` will force client-eval.

**Comment by jonathanantoine** [comment]
@bricelam I still have this issue on 2.2.4 : normal ? Here is a repro project : [TestProject1.zip](https://github.com/aspnet/EntityFrameworkCore/files/3084087/TestProject1.zip)

**Comment by bricelam** [comment]
@jonathanantoine Can you clarify? What are you trying and what do you expect to happen?

**Comment by jonathanantoine** [comment]
@bricelam I am trying to store decimals in a DbContext using a Sqlite driver and be able to order them :)

**Comment by bricelam** [comment]
The OrderBy should be evaluated on the client (i.e. in .NET). SQLite doesn't naively support 128-bit floating point operations.

**Comment by jonathanantoine** [comment]
I agree but as you can see in my repro project, it does not seems to be the case : or am I doing something wrong in my project setup ?

**Comment by bricelam** [comment]
...?
```
Test started...
 Summary: Passed: 1, Failed: 0, Ignored: 0
```

**Comment by jonathanantoine** [comment]
I am really sorry @bricelam : this was an issue with my IDE which didn't clear my cache and was still using the previous EF version.
